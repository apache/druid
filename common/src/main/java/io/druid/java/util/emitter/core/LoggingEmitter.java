/*
 * Licensed to Metamarkets Group Inc. (Metamarkets) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Metamarkets licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package io.druid.java.util.emitter.core;

/**
 */

import com.fasterxml.jackson.databind.ObjectMapper;
import io.druid.java.util.common.StringUtils;
import io.druid.java.util.common.lifecycle.LifecycleStart;
import io.druid.java.util.common.lifecycle.LifecycleStop;
import io.druid.java.util.common.logger.Logger;

import java.io.IOException;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 */
public class LoggingEmitter implements Emitter
{
  private final Logger log;
  private final Level level;
  private final ObjectMapper jsonMapper;

  private final AtomicBoolean started = new AtomicBoolean(false);

  public LoggingEmitter(LoggingEmitterConfig config, ObjectMapper jsonMapper)
  {
    this(new Logger(config.getLoggerClass()), Level.toLevel(config.getLogLevel()), jsonMapper);
  }

  public LoggingEmitter(Logger log, Level level, ObjectMapper jsonMapper)
  {
    this.log = log;
    this.level = level;
    this.jsonMapper = jsonMapper;
  }

  @Override
  @LifecycleStart
  public void start()
  {
    final boolean alreadyStarted = started.getAndSet(true);
    if (!alreadyStarted) {
      final String message = "Start: started [%s]";
      switch (level) {
        case TRACE:
          if (log.isTraceEnabled()) {
            log.trace(message, started.get());
          }
          break;
        case DEBUG:
          if (log.isDebugEnabled()) {
            log.debug(message, started.get());
          }
          break;
        case INFO:
          if (log.isInfoEnabled()) {
            log.info(message, started.get());
          }
          break;
        case WARN:
          log.warn(message, started.get());
          break;
        case ERROR:
          log.error(message, started.get());
          break;
      }
    }
  }

  @Override
  public void emit(Event event)
  {
    synchronized (started) {
      if (!started.get()) {
        throw new RejectedExecutionException("Service not started.");
      }
    }
    try {
      final String message = "Event [%s]";
      switch (level) {
        case TRACE:
          if (log.isTraceEnabled()) {
            log.trace(message, jsonMapper.writeValueAsString(event));
          }
          break;
        case DEBUG:
          if (log.isDebugEnabled()) {
            log.debug(message, jsonMapper.writeValueAsString(event));
          }
          break;
        case INFO:
          if (log.isInfoEnabled()) {
            log.info(message, jsonMapper.writeValueAsString(event));
          }
          break;
        case WARN:
          log.warn(message, jsonMapper.writeValueAsString(event));
          break;
        case ERROR:
          log.error(message, jsonMapper.writeValueAsString(event));
          break;
      }
    }
    catch (Exception e) {
      log.warn(e, "Failed to generate json");
    }
  }

  @Override
  public void flush() throws IOException
  {

  }

  @Override
  @LifecycleStop
  public void close() throws IOException
  {
    final boolean wasStarted = started.getAndSet(false);
    if (wasStarted) {
      final String message = "Close: started [%s]";
      switch (level) {
        case TRACE:
          if (log.isTraceEnabled()) {
            log.trace(message, started.get());
          }
          break;
        case DEBUG:
          if (log.isDebugEnabled()) {
            log.debug(message, started.get());
          }
          break;
        case INFO:
          if (log.isInfoEnabled()) {
            log.info(message, started.get());
          }
          break;
        case WARN:
          log.warn(message, started.get());
          break;
        case ERROR:
          log.error(message, started.get());
          break;
      }
    }
  }

  @Override
  public String toString()
  {
    return "LoggingEmitter{" +
           "log=" + log +
           ", level=" + level +
           '}';
  }

  public enum Level
  {
    TRACE,
    DEBUG,
    INFO,
    WARN,
    ERROR;

    public static Level toLevel(String name)
    {
      return Level.valueOf(StringUtils.toUpperCase(name));
    }
  }
}
