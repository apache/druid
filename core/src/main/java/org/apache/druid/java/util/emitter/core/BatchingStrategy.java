/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.druid.java.util.emitter.core;

public enum BatchingStrategy
{
  ARRAY {

    @Override
    public int batchStartLength()
    {
      return 1;
    }

    @Override
    public int separatorLength()
    {
      return 1;
    }

    @Override
    public int batchEndLength()
    {
      return 2;
    }

    @Override
    public int writeBatchStart(byte[] buffer)
    {
      buffer[0] = '[';
      return batchStartLength();
    }

    @Override
    public int writeMessageSeparator(byte[] buffer, int bufferOffset)
    {
      buffer[bufferOffset] = ',';
      return bufferOffset + separatorLength();
    }

    @Override
    public int writeBatchEnd(byte[] buffer, int bufferOffset)
    {
      buffer[bufferOffset] = ']';
      buffer[bufferOffset + 1] = '\n';
      return bufferOffset + batchEndLength();
    }
  },
  NEWLINES {

    @Override
    public int batchStartLength()
    {
      return 0;
    }

    @Override
    public int separatorLength()
    {
      return 1;
    }

    @Override
    public int batchEndLength()
    {
      return 1;
    }

    @Override
    public int writeBatchStart(byte[] buffer)
    {
      // Write nothing
      return batchStartLength();
    }

    @Override
    public int writeMessageSeparator(byte[] buffer, int bufferOffset)
    {
      buffer[bufferOffset] = '\n';
      return bufferOffset + separatorLength();
    }

    @Override
    public int writeBatchEnd(byte[] buffer, int bufferOffset)
    {
      return writeMessageSeparator(buffer, bufferOffset);
    }
  },
  ONLY_EVENTS {
    @Override
    public int batchStartLength()
    {
      return 0;
    }

    @Override
    public int separatorLength()
    {
      return 0;
    }

    @Override
    public int batchEndLength()
    {
      return 0;
    }

    @Override
    public int writeBatchStart(byte[] buffer)
    {
      return 0;
    }

    @Override
    public int writeMessageSeparator(byte[] buffer, int bufferOffset)
    {
      return bufferOffset;
    }

    @Override
    public int writeBatchEnd(byte[] buffer, int bufferOffset)
    {
      return bufferOffset;
    }
  };

  public abstract int batchStartLength();

  public abstract int separatorLength();

  public abstract int batchEndLength();

  public abstract int writeBatchStart(byte[] buffer);

  public abstract int writeMessageSeparator(byte[] buffer, int bufferOffset);

  public abstract int writeBatchEnd(byte[] buffer, int bufferOffset);
}
