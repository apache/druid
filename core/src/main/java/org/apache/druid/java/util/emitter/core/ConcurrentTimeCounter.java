/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.druid.java.util.emitter.core;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.primitives.UnsignedInts;

import javax.annotation.Nullable;
import java.util.concurrent.atomic.AtomicLong;

/**
 * A class to accumulate simple stats of some time points. All methods are safe to use from multiple threads.
 */
public class ConcurrentTimeCounter
{
  /** Lower 32 bits for sum of {@link #add}ed times, higher 32 bits for the count */
  private final AtomicLong timeSumAndCount = new AtomicLong(0L);
  /** Lower 32 bits for the max {@link #add}ed time, 63th bit for indication if any value is added. */
  private final AtomicLong max = new AtomicLong(-1);
  /** Similar to {@link #max} */
  private final AtomicLong min = new AtomicLong(-1);

  public void add(int time)
  {
    long x = (1L << 32) | time;
    timeSumAndCount.addAndGet(x);
    updateMax(time);
    updateMin(time);
  }

  private void updateMax(int time)
  {
    long max;
    do {
      max = this.max.get();
      if (max >= 0 && ((int) max) >= time) {
        return;
      }
    } while (!this.max.compareAndSet(max, UnsignedInts.toLong(time)));
  }

  private void updateMin(int time)
  {
    long min;
    do {
      min = this.min.get();
      if (min >= 0 && ((int) min) <= time) {
        return;
      }
    } while (!this.min.compareAndSet(min, UnsignedInts.toLong(time)));
  }

  @VisibleForTesting
  long getTimeSumAndCount()
  {
    return timeSumAndCount.get();
  }

  public long getTimeSumAndCountAndReset()
  {
    return timeSumAndCount.getAndSet(0L);
  }

  /**
   * Returns the max time {@link #add}ed since the previous call to this method or since the creation of this object,
   * or null if no times were added.
   */
  @Nullable
  public Integer getAndResetMaxTime()
  {
    long max = this.max.getAndSet(-1);
    if (max >= 0) {
      return (int) max;
    } else {
      return null;
    }
  }

  /**
   * Returns the min time {@link #add}ed since the previous call to this method or since the creation of this object,
   * or null if no times were added.
   */
  @Nullable
  public Integer getAndResetMinTime()
  {
    long min = this.min.getAndSet(-1);
    if (min >= 0) {
      return (int) min;
    } else {
      return null;
    }
  }

  public static int timeSum(long timeSumAndCount)
  {
    return (int) timeSumAndCount;
  }

  public static int count(long timeSumAndCount)
  {
    return (int) (timeSumAndCount >> 32);
  }
}
