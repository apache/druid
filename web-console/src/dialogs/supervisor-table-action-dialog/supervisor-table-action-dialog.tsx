/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';

import { ShowJson } from '../../components';
import { ShowHistory } from '../../components/show-history/show-history';
import { SupervisorStatisticsTable } from '../../components/supervisor-statistics-table/supervisor-statistics-table';
import { deepGet } from '../../utils';
import { BasicAction } from '../../utils/basic-action';
import { SideButtonMetaData, TableActionDialog } from '../table-action-dialog/table-action-dialog';

interface SupervisorTableActionDialogProps {
  supervisorId: string;
  actions: BasicAction[];
  onClose: () => void;
}

export const SupervisorTableActionDialog = React.memo(function SupervisorTableActionDialog(
  props: SupervisorTableActionDialogProps,
) {
  const { supervisorId, actions, onClose } = props;
  const [activeTab, setActiveTab] = useState('status');

  const supervisorTableSideButtonMetadata: SideButtonMetaData[] = [
    {
      icon: 'dashboard',
      text: 'Status',
      active: activeTab === 'status',
      onClick: () => setActiveTab('status'),
    },
    {
      icon: 'chart',
      text: 'Statistics',
      active: activeTab === 'stats',
      onClick: () => setActiveTab('stats'),
    },
    {
      icon: 'align-left',
      text: 'Payload',
      active: activeTab === 'payload',
      onClick: () => setActiveTab('payload'),
    },
    {
      icon: 'history',
      text: 'History',
      active: activeTab === 'history',
      onClick: () => setActiveTab('history'),
    },
  ];

  return (
    <TableActionDialog
      sideButtonMetadata={supervisorTableSideButtonMetadata}
      onClose={onClose}
      title={`Supervisor: ${supervisorId}`}
      actions={actions}
    >
      {activeTab === 'status' && (
        <ShowJson
          endpoint={`/druid/indexer/v1/supervisor/${supervisorId}/status`}
          transform={x => deepGet(x, 'payload')}
          downloadFilename={`supervisor-status-${supervisorId}.json`}
        />
      )}
      {activeTab === 'stats' && (
        <SupervisorStatisticsTable
          supervisorId={supervisorId}
          downloadFilename={`supervisor-stats-${supervisorId}.json`}
        />
      )}
      {activeTab === 'payload' && (
        <ShowJson
          endpoint={`/druid/indexer/v1/supervisor/${supervisorId}`}
          downloadFilename={`supervisor-payload-${supervisorId}.json`}
        />
      )}
      {activeTab === 'history' && (
        <ShowHistory
          endpoint={`/druid/indexer/v1/supervisor/${supervisorId}/history`}
          downloadFilename={`supervisor-history-${supervisorId}.json`}
        />
      )}
    </TableActionDialog>
  );
});
