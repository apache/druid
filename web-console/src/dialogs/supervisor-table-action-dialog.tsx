/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IDialogProps } from '@blueprintjs/core';
import * as React from 'react';

import { ShowJson } from '../components/show-json';
import { BasicAction, basicActionsToButtons } from '../utils/basic-action';

import { SideButtonMetaData, TableActionDialog } from './table-action-dialog';

interface SupervisorTableActionDialogProps extends IDialogProps {
  supervisorId: string;
  actions: BasicAction[];
  onClose: () => void;
}

interface SupervisorTableActionDialogState {
  activeTab: 'payload' | 'status' | 'stats' | 'history';
}

export class SupervisorTableActionDialog extends React.Component<SupervisorTableActionDialogProps, SupervisorTableActionDialogState> {
  constructor(props: SupervisorTableActionDialogProps) {
    super(props);
    this.state = {
      activeTab: 'payload'
    };
  }

  render(): React.ReactNode {
    const { supervisorId, actions, onClose } = this.props;
    const { activeTab } = this.state;

    const supervisorTableSideButtonMetadata: SideButtonMetaData[] = [
      {
        icon: 'align-left',
        text: 'Payload',
        active: activeTab === 'payload',
        onClick: () => this.setState({ activeTab: 'payload' })
      },
      {
        icon: 'dashboard',
        text: 'Status',
        active: activeTab === 'status',
        onClick: () => this.setState({ activeTab: 'status' })
      },
      {
        icon: 'chart',
        text: 'Statistics',
        active: activeTab === 'stats',
        onClick: () => this.setState({ activeTab: 'stats' })
      },
      {
        icon: 'history',
        text: 'History',
        active: activeTab === 'history',
        onClick: () => this.setState({ activeTab: 'history' })
      }
    ];

    return <TableActionDialog
      isOpen
      sideButtonMetadata={supervisorTableSideButtonMetadata}
      onClose={onClose}
      title={`Supervisor: ${supervisorId}`}
      bottomButtons={basicActionsToButtons(actions)}
    >
      {activeTab === 'payload' && <ShowJson endpoint={`/druid/indexer/v1/supervisor/${supervisorId}`} downloadFilename={`supervisor-payload-${supervisorId}.json`}/>}
      {activeTab === 'status' && <ShowJson endpoint={`/druid/indexer/v1/supervisor/${supervisorId}/status`} downloadFilename={`supervisor-status-${supervisorId}.json`}/>}
      {activeTab === 'stats' && <ShowJson endpoint={`/druid/indexer/v1/supervisor/${supervisorId}/stats`} downloadFilename={`supervisor-stats-${supervisorId}.json`}/>}
      {activeTab === 'history' && <ShowJson endpoint={`/druid/indexer/v1/supervisor/${supervisorId}/history`} downloadFilename={`supervisor-history-${supervisorId}.json`}/>}
    </TableActionDialog>;
  }
}
