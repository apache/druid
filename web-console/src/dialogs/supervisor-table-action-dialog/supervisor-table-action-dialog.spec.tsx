/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import {Intent} from '@blueprintjs/core';
import * as React from 'react';
import { render } from 'react-testing-library';

import {SupervisorTableActionDialog} from './supervisor-table-action-dialog';

const basicAction = {title: 'test', onAction: () => null};
describe('describe supervisor table action dialog', () => {
  it('supervisor table action dialog snapshot', () => {
    const supervisorTableActionDialog =
      <SupervisorTableActionDialog
        supervisorId={'test'}
        actions={[basicAction, basicAction, basicAction, basicAction]}
        onClose={() => null}
        isOpen={true}
      />;
    const { container, getByText } = render(supervisorTableActionDialog, { container: document.body });
    expect(container.firstChild).toMatchSnapshot();
  });
});
