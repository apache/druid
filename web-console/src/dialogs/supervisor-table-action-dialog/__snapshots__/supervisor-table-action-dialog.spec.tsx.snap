// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SupervisorTableActionDialog matches snapshot 1`] = `
<div
  class="bp5-portal"
>
  <div
    aria-live="polite"
    class="bp5-overlay bp5-overlay-open bp5-overlay-scroll-container"
  >
    <div
      class="bp5-overlay-start-focus-trap bp5-overlay-appear bp5-overlay-appear-active"
      tabindex="0"
    />
    <div
      class="bp5-overlay-backdrop bp5-overlay-appear bp5-overlay-appear-active"
    />
    <div
      class="bp5-dialog-container bp5-overlay-content bp5-overlay-appear bp5-overlay-appear-active"
      tabindex="0"
    >
      <div
        aria-labelledby="title-bp-dialog-0"
        aria-modal="true"
        class="bp5-dialog table-action-dialog"
        role="dialog"
      >
        <div
          class="bp5-dialog-header"
        >
          <h6
            class="bp5-heading"
            id="title-bp-dialog-0"
          >
            Supervisor: test
          </h6>
          <button
            aria-label="Close"
            class="bp5-button bp5-minimal bp5-dialog-close-button"
            type="button"
          >
            <span
              aria-hidden="true"
              class="bp5-icon bp5-icon-small-cross"
            >
              <svg
                data-icon="small-cross"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M188.2 160L234 205.8C237.8 209.4 240 214.4 240 220C240 231 231 240 220 240C214.4 240 209.4 237.8 205.8 234.2L160 188.2L114.2 234.2C110.6 237.8 105.6 240 100 240C89 240 80 231 80 220C80 214.4 82.2 209.4 85.8 205.8L131.8 160L86 114.2C82.2 110.6 80 105.6 80 100C80 89 89 80 100 80C105.6 80 110.6 82.2 114.2 85.8L160 131.8L205.8 86C209.4 82.2 214.4 80 220 80C231 80 240 89 240 100C240 105.6 237.8 110.6 234.2 114.2L188.2 160z"
                  fill-rule="evenodd"
                  style="transform-origin: center;"
                  transform="scale(0.05, -0.05) translate(-160, -160)"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="bp5-dialog-body"
        >
          <div
            class="side-bar"
          >
            <button
              class="bp5-button bp5-intent-primary tab-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="bp5-icon bp5-icon-dashboard"
              >
                <svg
                  data-icon="dashboard"
                  height="20"
                  role="img"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M6 5c-.55 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1zM4 9c-.55 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1zm6-4c.55 0 1-.45 1-1s-.45-1-1-1-1 .45-1 1 .45 1 1 1zm0-5C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm6-9c-.55 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1zm-8 5c0 1.1.9 2 2 2s2-.9 2-2c0-.33-2-8-2-8s-2 7.67-2 8zm6-9c-.55 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp5-button-text"
              >
                Status
              </span>
            </button>
            <button
              class="bp5-button bp5-minimal tab-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="bp5-icon bp5-icon-chart"
              >
                <svg
                  data-icon="chart"
                  height="20"
                  role="img"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M7 11v8c0 .55.45 1 1 1h4c.55 0 1-.45 1-1v-8l-2 2-4-2zm-7 8c0 .55.45 1 1 1h4c.55 0 1-.45 1-1v-8l-6 3v5zM17 7l-3 3v9c0 .55.45 1 1 1h4c.55 0 1-.45 1-1V8.74c-.26.15-.58.26-1 .26-1.92 0-2-2-2-2zm2-6h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59L10.8 8.78 7.45 7.11v.01C7.31 7.05 7.16 7 7 7s-.31.05-.44.11V7.1l-6 3v.01c-.33.17-.56.5-.56.89 0 .55.45 1 1 1 .16 0 .31-.05.44-.11v.01L7 9.12l3.55 1.78v-.01c.14.06.29.11.45.11.28 0 .53-.11.71-.29L18 4.41V6c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp5-button-text"
              >
                Task stats
              </span>
            </button>
            <button
              class="bp5-button bp5-minimal tab-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="bp5-icon bp5-icon-align-left"
              >
                <svg
                  data-icon="align-left"
                  height="20"
                  role="img"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M1 7h10c.55 0 1-.45 1-1s-.45-1-1-1H1c-.55 0-1 .45-1 1s.45 1 1 1zm0-4h18c.55 0 1-.45 1-1s-.45-1-1-1H1c-.55 0-1 .45-1 1s.45 1 1 1zm14 14H1c-.55 0-1 .45-1 1s.45 1 1 1h14c.55 0 1-.45 1-1s-.45-1-1-1zm4-8H1c-.55 0-1 .45-1 1s.45 1 1 1h18c.55 0 1-.45 1-1s-.45-1-1-1zM1 15h6c.55 0 1-.45 1-1s-.45-1-1-1H1c-.55 0-1 .45-1 1s.45 1 1 1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp5-button-text"
              >
                Spec
              </span>
            </button>
            <button
              class="bp5-button bp5-minimal tab-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="bp5-icon bp5-icon-history"
              >
                <svg
                  data-icon="history"
                  height="20"
                  role="img"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M10 0C6.71 0 3.82 1.6 2 4.05V2c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.76C5.23 3.17 7.47 2 10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8c0-.55-.45-1-1-1s-1 .45-1 1c0 5.52 4.48 10 10 10s10-4.48 10-10S15.52 0 10 0zm0 3c-.55 0-1 .45-1 1v6c0 .28.11.53.29.71l3 3a1.003 1.003 0 001.42-1.42L11 9.59V4c0-.55-.45-1-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp5-button-text"
              >
                History
              </span>
            </button>
          </div>
          <div
            class="main-section"
          >
            <div
              class="show-json"
            >
              <div
                class="top-actions"
              >
                <div
                  class="bp5-button-group right-buttons"
                >
                  <button
                    class="bp5-button bp5-disabled bp5-minimal"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="bp5-button-text"
                    >
                      Download
                    </span>
                  </button>
                  <button
                    class="bp5-button bp5-disabled bp5-minimal"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="bp5-button-text"
                    >
                      Copy
                    </span>
                  </button>
                  <button
                    class="bp5-button bp5-disabled bp5-minimal"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="bp5-button-text"
                    >
                      View raw
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="main-area"
              >
                <div
                  class="loader"
                >
                  <div
                    class="loader-logo"
                  >
                    <svg
                      viewBox="0 0 100 100"
                    >
                      <path
                        class="one"
                        d="M54.2,69.8h-2.7c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h2.7c11.5,0,23.8-7.4,23.8-23.7
          c0-9.1-6.9-15.8-16.4-15.8H38c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h23.6c5.3,0,10.1,1.9,13.6,5.3c3.5,3.4,5.4,8,5.4,13.1
          c0,6.6-2.3,13-6.3,17.7C69.5,66.8,62.5,69.8,54.2,69.8z"
                      />
                      <path
                        class="two"
                        d="M55.7,59.5h-26c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h26c7.5,0,11.5-5.8,11.5-11.5
            c0-4.2-3.2-7.3-7.7-7.3h-26c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h26c5.9,0,10.3,4.3,10.3,9.9c0,3.7-1.3,7.2-3.7,9.8
            C63.5,58,59.9,59.5,55.7,59.5z"
                      />
                      <path
                        class="three"
                        d="M27.2,38h-6.3c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h6.3c0.7,0,1.3,0.6,1.3,1.3S27.9,38,27.2,38z"
                      />
                      <path
                        class="four"
                        d="M45.1,69.8h-5.8c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h5.8c0.7,0,1.3,0.6,1.3,1.3
            C46.4,69.2,45.8,69.8,45.1,69.8z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="bp5-dialog-footer"
        >
          <div
            class="footer-actions-left"
          >
            <span
              class="bp5-popover-target"
            >
              <button
                aria-expanded="false"
                aria-haspopup="menu"
                class="bp5-button"
                type="button"
              >
                <span
                  aria-hidden="true"
                  class="bp5-icon bp5-icon-wrench"
                >
                  <svg
                    data-icon="wrench"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M15.83 3.7l-3.06 3.05-2.84-.7-.7-2.83L12.29.17a5.004 5.004 0 00-4.83 1.29 4.967 4.967 0 00-1.12 5.36L.58 12.58c-.36.36-.58.86-.58 1.41 0 1.1.9 2 2 2 .55 0 1.05-.22 1.41-.59l5.77-5.77c1.79.69 3.91.33 5.35-1.12 1.32-1.3 1.74-3.15 1.3-4.81z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp5-button-text"
                >
                  Actions
                </span>
                <span
                  aria-hidden="true"
                  class="bp5-icon bp5-icon-caret-down"
                >
                  <svg
                    data-icon="caret-down"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </div>
          <div
            class="bp5-dialog-footer-actions"
          >
            <button
              class="bp5-button bp5-intent-primary"
              type="button"
            >
              <span
                class="bp5-button-text"
              >
                Close
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="bp5-overlay-end-focus-trap bp5-overlay-appear bp5-overlay-appear-active"
      tabindex="0"
    />
  </div>
</div>
`;
